### PR TITLE
Batched mitigate

### LINKED ISSUE #1253: Make mitigate_executor compatible with batched executor functions
### Description
All error mitigation techniques in Mitiq have the associated `mitigate_executor` function. E.g. `mitiq.zne.mitigate_executor()`.
The current implementation for ZNE, PEC and CDR assumes an executor that is not batched, i.e., that takes a single circuit
and returns a single result.
Only for the DDD technique, `mitigate_executor` is now compatible with batched executors.

 ### How to fix
- Apply a solution similar to the code of `mitiq.ddd.mitigate_executor()` to `mitiq.[xxx].mitigate_executor()` for `xxx in [zne, pec, cdr]`.
- Add a test for each technique to make sure it works.

### PR DESCRIPTION
**Pull request by nylewong**
Description
-----------

Fixes #1253 

Completed support for batched mitigate in ZNE, PEC, and CDR. 

Checklist
---------

Check off the following once complete (or if not applicable) after opening the PR. The PR will be reviewed once this checklist is complete and all tests are passing.

- [x] I added unit tests for new code.
- [x] I used [type hints](https://www.python.org/dev/peps/pep-0484/) in function signatures.
- [x] I used [Google-style](https://sphinxcontrib-napoleon.readthedocs.io/en/latest/example_google.html) docstrings for functions.
- [x] I [updated the documentation](../blob/master/docs/CONTRIBUTING_DOCS.md) where relevant.
- [ ] Added myself / the copyright holder to the AUTHORS file

If some items remain, you can mark this a [draft pull request](https://github.blog/2019-02-14-introducing-draft-pull-requests/).

License
-------

- [ ] I license this contribution under the terms of the GNU GPL, version 3 and grant Unitary Fund the right to provide additional permissions as described in section 7 of the GNU GPL, version 3.

Tips
----

- If the validation check fails:

    1. Run `make check-types` (from the root directory of the repository) and fix any [mypy](https://mypy.readthedocs.io/en/stable/) errors.

    2. Run `make check-style` and fix any [flake8](http://flake8.pycqa.org) errors.

    3. Run `make format` to format your code with the [black](https://black.readthedocs.io/en/stable/index.html) autoformatter.

  For more information, check the [Mitiq style guidelines](https://mitiq.readthedocs.io/en/stable/contributing.html#style-guidelines).

- Write "Fixes #XYZ" in the description if this PR fixes Issue #XYZ.

### CODE REVIEW

**Comment by Misty-W** [comment]
hi @nylewong, thanks for the PR! Looks like you're off to a good start on adding support for the batched executor in ZNE, PEC, and CDR.

Can you remove the folder with your environment files from the PR? They shouldn't be needed for the solution to the issue, and it would make it much easier to load and review the diff. Feel free to ping here or on Discord if you need help.

**Comment by Misty-W** [comment]
hi @nylewong , hope you can join the community call tomorrow and build some momentum on this PR!

Either way, I think the best way forward is to revert the commit you made with the environment files, since they aren't needed and make it difficult to review the relevant Mitiq files. Let me know if you need help with this and I can walk you through!

File: mitiq/cdr/tests/test_cdr.py [hunk 1]
```diff
@@ -41,6 +43,8 @@
 def execute(circuit: QPROGRAM) -> np.ndarray:
     return compute_density_matrix(convert_to_mitiq(circuit)[0])
 
+def batched_execute(circuits) -> List[np.ndarray]:
+    return [execute(circuit) for circuit for circuits]
```
Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:47]:
```suggestion
    return [execute(circuit) for circuit in circuits]
```
```suggestion
    return [execute(circuit) for circuit for circuits]
```

Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:47]:
```suggestion
    return [execute(circuit) for circuit in circuits]
```
```suggestion
    return [execute(circuit) for circuit for circuits]
```

**Comment by Misty-W** [comment]
hi @nylewong, thanks for updating the PR! Can you add a test to `test_zne.py` like you did for the other techniques?

**Comment by Misty-W** [comment]
hi @nylewong, I'm happy to help with any questions or blockers you might have!

File: mitiq/ddd/tests/test_ddd.py [hunk 2]
```diff
@@ -181,6 +181,9 @@ def test_mitigate_executor_ddd():
     mitigated_executor = mitigate_executor(serial_executor, rule=xx)
     assert np.isclose(mitigated_executor(circuit_cirq_a), ddd_value)
 
+    batched_mitigated_executor = mitigate_executor(batched_executor, rule=xx)
+    assert np.isclose(*batched_mitigated_executor([circuit_cirq_a]))
```
Comment (Misty-W) [inline mitiq/ddd/tests/test_ddd.py:185]:
```suggestion
    assert np.isclose(*batched_mitigated_executor([circuit_cirq_a]), ddd_value)
```

File: mitiq/pec/tests/test_pec.py [hunk 3]
```diff
@@ -493,6 +493,18 @@ def test_mitigate_executor_qiskit():
 
     assert np.isclose(unmitigated, mitigated)
 
+    batched_unmitigated = batched_executor([circuit])
+
+    batched_mitigated_executor = mitigated_executor(
```
Comment (Misty-W) [inline mitiq/pec/tests/test_pec.py:498]:
```suggestion
    batched_mitigated_executor = mitigate_executor(
```

File: mitiq/ddd/tests/test_ddd.py [hunk 4]
```diff
@@ -168,7 +168,7 @@ def test_execute_with_ddd_with_full_output():
     assert len(ddd_data["circuits_with_ddd"]) == 2
     assert len(ddd_data["ddd_trials"]) == 2
     assert ddd_data["ddd_value"] == ddd_value
-    # For a deterministic rule
+    # For a deterministic rulecs231n
```
Comment (Misty-W) [inline mitiq/ddd/tests/test_ddd.py:171]:
```suggestion
    # For a deterministic rule
```

File: mitiq/zne/tests/test_zne.py [hunk 5]
```diff
@@ -321,7 +325,11 @@ def test_qiskit_mitigate_executor():
     mitigated_executor = mitigate_executor(qiskit_executor)
     zne_value = mitigated_executor(circuit)
     assert abs(true_zne_value - zne_value) < abs(true_zne_value - base)
-
+
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:328]:
```suggestion

```

File: mitiq/zne/tests/test_zne.py [hunk 6]
```diff
@@ -321,7 +325,11 @@ def test_qiskit_mitigate_executor():
     mitigated_executor = mitigate_executor(qiskit_executor)
     zne_value = mitigated_executor(circuit)
     assert abs(true_zne_value - zne_value) < abs(true_zne_value - base)
-
+    
+    batched_mitigated_executor = mitigate_executor(batched_qiskit_executor)
+    batched_zne_value = batched_mitigated_executor([circuit])
+    assert abs(true_zne_value - batched_zne_value) < abs(true_zne_value - base)
+
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:332]:
```suggestion

```

File: mitiq/zne/tests/test_zne.py [hunk 7]
```diff
@@ -321,7 +325,9 @@ def test_qiskit_mitigate_executor():
     mitigated_executor = mitigate_executor(qiskit_executor)
     zne_value = mitigated_executor(circuit)
     assert abs(true_zne_value - zne_value) < abs(true_zne_value - base)
-
+    batched_mitigated_executor = mitigate_executor(batched_qiskit_executor)
+    batched_zne_value = batched_mitigated_executor([circuit])
+    assert abs(true_zne_value - batched_zne_value) < abs(true_zne_value - base)
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:331]:
```suggestion


```

File: mitiq/zne/tests/test_zne.py [hunk 8]
```diff
@@ -304,6 +305,10 @@ def qiskit_decorated_executor(qp: QPROGRAM) -> float:
     return qiskit_executor(qp)
 
 
+def batched_qiskit_execute(circuits) -> List[float]:
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:308]:
```suggestion
def batched_qiskit_executor(circuits) -> List[float]:
```

File: mitiq/zne/tests/test_zne.py [hunk 9]
```diff
@@ -321,6 +326,9 @@ def test_qiskit_mitigate_executor():
     mitigated_executor = mitigate_executor(qiskit_executor)
     zne_value = mitigated_executor(circuit)
     assert abs(true_zne_value - zne_value) < abs(true_zne_value - base)
+    batched_mitigated_executor = mitigate_executor(batched_qiskit_executor)
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:329]:
```suggestion
        batched_mitigated_executor = mitigate_executor(batched_qiskit_executor)
    batched_zne_values = batched_mitigated_executor([circuit])
    assert [
        abs(true_zne_value - batched_zne_value) < abs(true_zne_value - base)
        for batched_zne_value in batched_zne_values
    ]
```

File: mitiq/zne/tests/test_zne.py [hunk 10]
```diff
@@ -321,6 +326,9 @@ def test_qiskit_mitigate_executor():
...
+    batched_zne_value = batched_mitigated_executor([circuit])
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:330]:
```suggestion
    batched_zne_values = batched_mitigated_executor([circuit])
```

File: mitiq/zne/tests/test_zne.py [hunk 11]
```diff
@@ -321,6 +326,9 @@ def test_qiskit_mitigate_executor():
...
+    assert abs(true_zne_value - batched_zne_value) < abs(true_zne_value - base)
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:331]:
```suggestion
    assert [
        abs(true_zne_value - batched_zne_value) < abs(true_zne_value - base)
        for batched_zne_value in batched_zne_values
    ]
```

**Comment by Misty-W** [comment]
hi @natestemen, I picked this up since there hasn't been activity from the original contributor in a while. 

I'd say it's lower prio than the other issues open for the milestone but unless major changes are required, it'd be nice to have this in too.

**Comment by Misty-W** [comment]
> Great, @Misty-W! Is this ready for review? If it does not make it in this milestone I'd definitely put it in the next one.

@nathanshammah yes ready for review and requested @natestemen as reviewer.

File: mitiq/ddd/tests/test_ddd.py [hunk 12]
```diff
@@ -181,6 +181,9 @@ def test_mitigate_executor_ddd():
     mitigated_executor = mitigate_executor(serial_executor, rule=xx)
     assert np.isclose(mitigated_executor(circuit_cirq_a), ddd_value)
 
+    batched_mitigated_executor = mitigate_executor(batched_executor, rule=xx)
+    assert np.isclose(*batched_mitigated_executor([circuit_cirq_a]), ddd_value)
+
```
Comment (natestemen) [inline mitiq/ddd/tests/test_ddd.py:186]: I think we'd want at least two circuits here, right?

File: mitiq/pec/tests/test_pec.py [hunk 13]
```diff
@@ -493,6 +493,18 @@ def test_mitigate_executor_qiskit():
 
     assert np.isclose(unmitigated, mitigated)
 
+    batched_unmitigated = batched_executor([circuit])
```
Comment (natestemen) [inline mitiq/pec/tests/test_pec.py:496]: Again, is one circuit enough to fully test this?

**Review by natestemen — APPROVED** [review]
Nice work! It pretty much all looks good to me, but I do think we might want to ensure the tests pass when someone is passing a list of more than one `QPROGRAM` into more than one circuit into `mitigate_executor`. I'm worried a list with one item might not test the same things as a list with two, or many items.

**Question unrelated to PR:**
I noticed you used the type signatures from `execute_with_xxx` (where `xxx` is a method) when defining the `new_executor` which seems correct to me, but I'm curious why `execute_with_ddd` returns such a different type than all the other methods.

File: mitiq/pec/tests/test_pec.py (hunk 13 above)
Comment (Misty-W) [reply mitiq/pec/tests/test_pec.py:496]: Good point, I remember thinking it was a bit odd when I first reviewed and then forgot to change it.

File: mitiq/cdr/tests/test_cdr.py [hunk 14]
```diff
@@ -198,6 +204,21 @@ def test_mitigated_execute_with_cdr(
...
+        executor=batched_execute,
+        observable=obs,
+        simulator=simulate,
+        num_training_circuits=20,
+        fraction_non_clifford=0.5,
+        fit_function=fit_function,
+        random_state=random_state,
+        **kwargs,
+    )
+    cdr_batched_mitigated = cdr_batched_executor([circuit])
```
Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:217]:
```suggestion
    cdr_batched_mitigated = cdr_batched_executor([circuit] * 3)
```

File: mitiq/ddd/tests/test_ddd.py (hunk 12 above)
Comment (Misty-W) [reply mitiq/ddd/tests/test_ddd.py:186]:
```suggestion
    batched_mitigated_executor = mitigate_executor(batched_executor, rule=xx)
    assert np.isclose(*batched_mitigated_executor([circuit_cirq_a] * 3), ddd_value)

```

File: mitiq/pec/tests/test_pec.py [hunk 15]
```diff
@@ -493,6 +493,18 @@ def test_mitigate_executor_qiskit():
...
+
+    batched_mitigated_executor = mitigate_executor(
+        batched_executor,
+        representations=[rep],
+        num_samples=10,
+        random_state=1,
+    )
+    batched_mitigated = batched_mitigated_executor([circuit])
```
Comment (Misty-W) [inline mitiq/pec/tests/test_pec.py:504]:
```suggestion
    batched_mitigated = batched_mitigated_executor([circuit] * 3)
```

File: mitiq/zne/tests/test_zne.py [hunk 16]
```diff
@@ -321,6 +326,12 @@ def test_qiskit_mitigate_executor():
     mitigated_executor = mitigate_executor(qiskit_executor)
     zne_value = mitigated_executor(circuit)
     assert abs(true_zne_value - zne_value) < abs(true_zne_value - base)
+    batched_mitigated_executor = mitigate_executor(batched_qiskit_executor)
+    batched_zne_values = batched_mitigated_executor([circuit])
```
Comment (Misty-W) [inline mitiq/zne/tests/test_zne.py:330]:
```suggestion
    batched_zne_values = batched_mitigated_executor([circuit] *3)
```

File: mitiq/pec/tests/test_pec.py (hunk 13 above)
Comment (Misty-W) [inline mitiq/pec/tests/test_pec.py:496]:
```suggestion
    batched_unmitigated = batched_executor([circuit] * 3)
```

File: mitiq/cdr/tests/test_cdr.py [hunk 17]
```diff
@@ -198,6 +204,21 @@ def test_mitigated_execute_with_cdr(
...
+    assert abs(cdr_batched_mitigated - true_value) <= abs(
```
Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:218]:
```suggestion
    assert [abs(cdr_batched_mitigated - true_value) <= abs(
```

File: mitiq/cdr/tests/test_cdr.py [hunk 18]
```diff
@@ -198,6 +204,21 @@ def test_mitigated_execute_with_cdr(
...
+        noisy_value - true_value
+    )
```
Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:220]:
```suggestion
    assert [
        abs(cdr_batched_mitigated - true_value) <= abs(noisy_value - true_value)
        for cdr_batched_mitigated in cdr_batched_mitigated_values
    ]    
```

File: mitiq/ddd/tests/test_ddd.py [hunk 19]
```diff
@@ -181,6 +181,9 @@ def test_mitigate_executor_ddd():
     mitigated_executor = mitigate_executor(serial_executor, rule=xx)
     assert np.isclose(mitigated_executor(circuit_cirq_a), ddd_value)
 
+    batched_mitigated_executor = mitigate_executor(batched_executor, rule=xx)
+    assert np.isclose(*batched_mitigated_executor([circuit_cirq_a] * 3), ddd_value)
```
Comment (Misty-W) [inline mitiq/ddd/tests/test_ddd.py:185]:
```suggestion
    assert np.isclose(
        *batched_mitigated_executor([circuit_cirq_a] * 3), ddd_value
    )```

File: mitiq/cdr/tests/test_cdr.py [hunk 20]
```diff
@@ -198,6 +204,22 @@ def test_mitigated_execute_with_cdr(
...
+        fraction_non_clifford=0.5,
+        fit_function=fit_function,
+        random_state=random_state,
+        **kwargs,
+    )
+    cdr_batched_mitigated_values = cdr_batched_executor([circuit] * 3)
+    assert [
+        abs(cdr_batched_mitigated - true_value) <= abs(noisy_value - true_value)
+        for cdr_batched_mitigated in cdr_batched_mitigated_values
+    ]
```
Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:221]:
```suggestion
    ]
```

Comment (Misty-W) [inline mitiq/cdr/tests/test_cdr.py:221]:
```suggestion
    assert [
        abs(cdr_batched_mitigated - true_value)
        <= abs(noisy_value - true_value)
        for cdr_batched_mitigated in cdr_batched_mitigated_values
    ]
```

File: mitiq/pec/tests/test_pec.py [hunk 21]
```diff
@@ -493,6 +493,18 @@ def test_mitigate_executor_qiskit():
...
+
+    batched_mitigated_executor = mitigate_executor(
+        batched_executor,
+        representations=[rep],
+        num_samples=10,
+        random_state=1,
+    )
+    batched_mitigated = batched_mitigated_executor([circuit] * 3)
+
+    assert np.isclose(batched_unmitigated, batched_mitigated)
```
Comment (Misty-W) [inline mitiq/pec/tests/test_pec.py:506]:
```suggestion
    assert [
        np.isclose(batched_unmitigated_value, batched_mitigated_value)
        for batched_unmitigated_value, batched_mitigated_value in zip(
            batched_unmitigated, batched_mitigated
        )
    ]
```

**Pull request merged**